codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -496,19 +520,49 @@ private void incrementPendingBytes(int numBytes) {
              */
             int write(int allowedBytes) {
                 int before = payload.size();
-                needFlush |= payload.write(Math.max(0, allowedBytes));
-                int writtenBytes = before - payload.size();
+                int writtenBytes = 0;
                 try {
-                    connectionState().incrementStreamWindow(-writtenBytes);
-                    incrementStreamWindow(-writtenBytes);
+                    boolean writeError = true;
+                    try {
+                        writing = true;
+                        needFlush |= payload.write(Math.max(0, allowedBytes));
+                        writing = false;
+                        writeError = false;
+                    } catch (Throwable t) {
+                        writing = false;
+                        pendingWriteQueue.remove();
+                        payload.error(t);
+                    }
+                    if (!writeError) {
```

Comment:
It seems we're using 3 flags here when we could use 2.  I propose you just have the 2 members `cancelled` and the `writing`.

`Cancelled` is always set in a call to `state.cancel()` and can be called multiple times.  When called it always immediately sets `cancelled` to true and clears/errors the pending queue if `wriitng == false`.  This allows us to call `cancel` at any any point in order to just set the flag.

So I think the method could then look something like this:

```java
            int write(int allowedBytes) {
                int before = payload.size();
                int writtenBytes = 0;
                try {
                    if (writing) {
                        throw new IllegalStateException("write is not re-entrant");
                    }
                    // Write the portion of the frame. 
                    writing = true;
                    needFlush |= payload.write(Math.max(0, allowedBytes));
                    if (!cancelled && payload.size() == 0) {
                        // This frame has been fully written, remove this frame
                        // and notify the payload.
                        pendingWriteQueue.remove();
                        payload.writeComplete();
                    }
                } catch (Exception t) {
                    cancel();
                } finally {
                    // Make sure we always decrement the flow control windows
                    // by the bytes written.
                    writtenBytes = before - payload.size();
                    decrementFlowControlWindow(writtenBytes);
                    decrementPendingBytes(writtenBytes);

                    // Indicate that we're now done writing. If a cancellation
                    // occurred while writing, call cancel again to clear and
                    // error all of the pending writes.
                    writing = false;
                    if (cancelled) {
                        cancel();
                    }
                }
                return writtenBytes;
            }
```